### PR TITLE
fix: Enhace black efficiently to  skip directories listed in .gitignore

### PR DESCRIPTION
**Pull request by Mr-Sunglasses**
### Description

fix: #4405  Enhace black efficiently to  skip directories listed in .gitignore

### Checklist - did you ...

- [ ] Add an entry in `CHANGES.md` if necessary?
- [X] Add / update tests if necessary?
- [ ] Add new / update outdated documentation?

### CODE REVIEW

File: src/black/files.py [hunk 1]
```diff
@@ -309,6 +309,8 @@ def _path_is_ignored(
     for gitignore_path, pattern in gitignore_dict.items():
         try:
             relative_path = path.relative_to(gitignore_path).as_posix()
+            if path.is_dir():
```
Comment (JelleZijlstra) [inline src/black/files.py:312]: Any way we could test that this works as expected?

Comment (Mr-Sunglasses) [reply src/black/files.py:312]: I've add the test for it

**Comment by Mr-Sunglasses** [comment]
@JelleZijlstra I've added the test for this and now black is almost 2x faster when it have to ignore directories from `.gitignore`

A benchmark I've tested, with `node_modules` as ignored.

Screenshot 2024-07-27 at 4 13 07 AM

**Comment by Mr-Sunglasses** [comment]
> @JelleZijlstra I've added the test for this and now black is almost 2x faster when it have to ignore directories from `.gitignore`
> 
> A benchmark I've tested, with `node_modules` as ignored.
> 
> Screenshot 2024-07-27 at 4 13 07 AM

A question, should we add a changelog entry as it's a simple but significant change?

**Comment by Mr-Sunglasses** [comment]
Should we add a changelog entry as it's a simple but significant change?

**Review by JelleZijlstra — APPROVED** [review]
Thank you! I pushed a news entry.

**Comment by hauntsaninja** [comment]
Note this probably hurts perf in the normal case due to adding a bunch of new stats
This is more correct, but could be nice to rework things to use os.scandir and away from pathlib.Path

**Pull request merged**